src/Common/src/CoreWCF/Runtime/NonReentrantAsyncLock.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CoreWCF.Runtime
+{
+    /// <summary>
+    /// This class provides a locking mechanism compatible with task async patterns.
+    /// </summary>
+    /// <remarks>
+    /// Unlike the <see cref="ReentrantAsyncLock"/> it does not provide a reentrant (aka. recursive) locking
+    /// and can therefore only be used in scenarios where we have clear control over the execution path.
+    /// </remarks>
+    internal sealed class NonReentrantAsyncLock
+    {
+        // based of the ideas in https://devblogs.microsoft.com/pfxteam/building-async-coordination-primitives-part-6-asynclock/
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+        private readonly Task<IDisposable> _releaser;
+#if DEBUG
+        private static readonly AsyncLocal<bool> s_reentranceFlag = new AsyncLocal<bool>();
+#endif
+
+        public NonReentrantAsyncLock()
+        {
+            _releaser = Task.FromResult<IDisposable>(new ReusableReleaser(this));
+        }
+
+        public Task<IDisposable> TakeLockAsync()
+        {
+#if DEBUG
+            Fx.Assert(!s_reentranceFlag.Value, "Usage of lock is reentrant but this class does not support it!");
+            s_reentranceFlag.Value = true;
+#endif
+            // we avoid allocation of new releaser instances if task does not need awaiting
+            // SemaphoreSlim might return a completed task if no actual waiting needed to be done.
+            var wait = _semaphore.WaitAsync();
+            return wait.IsCompleted
+                ? _releaser
+                : wait.ContinueWith(
+                    (_, state) => (IDisposable)new OneTimeReleaser((NonReentrantAsyncLock)state),
+                    this,
+                    CancellationToken.None,
+                    TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default);
+        }
+
+        /// <summary>
+        /// This releaser does a release on every dispose.
+        /// </summary>
+        private readonly struct ReusableReleaser : IDisposable
```

Comment:
I am a bit unsure about which path is best to follow. The remark of @StevenBonePgh to avoid allocations somehow requires us to have a disposable which does a release every time because this instance is reused for multiple unlocks. 

This opens again the path that some outside caller could also make a mistake and do too many releases which was raised as concern by @mconnew . The `OneTimeReleaser` would handle the single release for cases where we really did wait, but I guess it will not be used most of the time.

I think the mix we have right now is not good and we should clean this to one of the following options:

a) We keep the "ResusableReleaser" and use it everywhere. We neglect the risk that somebody misuses the lock, and by this gain the benefit that we can provide the same releaser instance in case no wait was done. 

b) We keep the "OneTimeReleaser" and use it everywhere. We always create new releaser instances and by this remove the benefit of saving some struct allocations. By this we keep the benefit that we prevent wrong usage of the Releasers.

Reply:
It might be worth returning a ValueTask. [Here's a good article](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/) by the designer of ValueTask about it. Reusable is a bit of a misnomer with a struct. It gets copied by value. So on line 28 ReusableReleaser is created on the local stack. When Task.FromResult is called, the stack pointer gets moved to create room for the calling parameters, including space for the entire struct. It then gets copied into that space on the stack. In the FromResult method, it then needs to box the struct as the field in the Task class which stores it is a reference type (IDisposable). With your current api you might as well leave it as a class to begin with as it's going to get boxed anyway. As the lock is internal, you could save the allocation by leaving it as a struct and making it internal, then defining TakeLockAsync to return Task<LockReleaser>. You wouldn't need to have a cached _releaser any more as there would be zero allocation. The using statement doesn't need an IDisposable, it needs an object which implements IDisposable. I just verified that the using statement doesn't cast to IDisposable so there wouldn't be any boxing. Returning a ValueTask provides the zero allocation for the no-wait scenario. Also I personally prefer when avoiding making the method async to place the alternative code into a separate method which is async. So something like this:
```c#
        public ValueTask<LockReleaser> TakeLockAsync()
        {
            var wait = _semaphore.WaitAsync();
            return wait.IsCompleted
                ? new ValueTask<LockReleaser>(new LockReleaser(this))
                : new ValueTask<LockReleaser>(InnerTakeLockAsync(wait);
        }
        public async Task<LockReleaser> InnerTakeLockAsync(Task wait)
        {
            await wait;
            return new LockReleaser(this);
        }
```
This way the only allocation is the _semaphore.WaitAsync. But you can do even better than that!! You can utilize the synchronous Wait method passing a timeout of 0ms. This will return immediately telling you whether it acquired the lock or if it couldn't, with zero waiting. So you can do this:
```c#
        public ValueTask<LockReleaser> TakeLockAsync()
        {
            bool acquired = _semaphore.Wait(0);
            return acquired 
                ? new ValueTask<LockReleaser>(new LockReleaser(this))
                : new ValueTask<LockReleaser>(InnerTakeLockAsync());
        }
        public async Task<LockReleaser> InnerTakeLockAsync()
        {
            await _semaphore.WaitAsync();
            return new LockReleaser(this);
        }
```
This is truly zero allocation when the lock is available.